### ISSUE TITLE
`torch.distributed.broadcast` silently fails if incorrect group is given

### ISSUE THREAD

**Issue by qysnn** [pytorch/pytorch#67600]
Code:
```
import torch
import torch.multiprocessing as mp
import torch.distributed as dist
import os


def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '17777'
    # initialize the process group
    dist.init_process_group("nccl", rank=rank, world_size=world_size)


def subprocess_fn(rank):
    setup(rank, 4)
    test_tensor = torch.Tensor([rank]).cuda()
    print(f'Before broadcast data on rank {rank} is {test_tensor}.')
    incorrect_group = dist.new_group([2, 3])
    if rank < 2:
        dist.broadcast(test_tensor, 1, group=incorrect_group)
    print(f'After broadcast data on rank {rank} is {test_tensor}.')


if __name__ == '__main__':
    mp.spawn(subprocess_fn, args=(), nprocs=4)

```
Output:
```
Before broadcast data on rank 0 is tensor([0.], device='cuda:0').
Before broadcast data on rank 2 is tensor([2.], device='cuda:0').
Before broadcast data on rank 3 is tensor([3.], device='cuda:0').
Before broadcast data on rank 1 is tensor([1.], device='cuda:0').
After broadcast data on rank 1 is tensor([1.], device='cuda:0').
After broadcast data on rank 3 is tensor([3.], device='cuda:0').
After broadcast data on rank 0 is tensor([0.], device='cuda:0').
```

As this code snippet shows, `torch.distributed.broadcast` silently fails when given incorrect group, even if the source rank is not in that group. I don't this behavior is documented. For large projects where the groups are initialized and used at different places, the only thing users observe is the inconsistency between ranks and it is hard to debug the root cause. Maybe there should be a warning in the documentation to explicitly warn users that `torch.distributed.broadcast` would do nothing if given incorrect group? Or preferably throw an exception when source or destination ranks are not in the parallel group? Thanks!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

**Comment by mrshenli** [comment]
Hey @qysnn, thanks for reporting this!

> Or preferably throw an exception when source or destination ranks are not in the parallel group?

I agree, this shouldn't be too hard. Let me try a fix.

**Comment by mrshenli** [comment]
First attempt: #67639

We cannot directly error out on non-member group instances, as that would break BC. #67639 is just trying to raise an warning message to help improve debuggability. 

Hey @qysnn, is this sufficient for your use case, or do you need hard error? If it's the latter, we can start a poll issue to collect user feedback on this behavior change.

**Comment by qysnn** [comment]
@mrshenli Thanks a lot for the quick fix! I think this warning would make debugging much easier and it is sufficient for my use case.

**Comment by mrshenli** [comment]
closed by #67639
